src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala

```diff
@@ -0,0 +1,255 @@
+package akka.persistence.cassandra.query
+
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.time.temporal.ChronoUnit
+import java.util.UUID
+
+import scala.concurrent.duration._
+
+import akka.actor.ActorLogging
+import akka.actor.DeadLetterSuppression
+import akka.actor.NoSerializationVerificationNeeded
+import akka.actor.OneForOneStrategy
+import akka.actor.Props
+import akka.actor.SupervisorStrategy
+import akka.persistence.cassandra.journal.CassandraJournal
+import akka.stream.actor.ActorPublisher
+import akka.stream.actor.ActorPublisherMessage.Cancel
+import akka.stream.actor.ActorPublisherMessage.Request
+import com.datastax.driver.core.PreparedStatement
+import com.datastax.driver.core.Session
+import com.datastax.driver.core.utils.UUIDs
+
+private[query] object EventsByTagPublisher {
+
+  def props(tag: String, fromOffset: UUID, refreshInterval: Option[FiniteDuration], settings: CassandraReadJournalConfig,
+            session: Session, preparedSelect: PreparedStatement): Props = {
+    refreshInterval match {
+      case Some(interval) ⇒
+        Props(new LiveEventsByTagPublisher(tag, fromOffset, interval,
+          settings, session, preparedSelect))
+      case None ⇒
+        Props(new CurrentEventsByTagPublisher(tag, fromOffset, settings, session, preparedSelect))
+    }
+  }
+
+  private[query] case object Continue extends DeadLetterSuppression
+
+  private[query] case class ReplayDone(count: Int, seqNumbers: SequenceNumbers)
+    extends DeadLetterSuppression
+  private[query] case class ReplayAborted(
+    seqNumbers: SequenceNumbers, persistenceId: String, expectedSeqNr: Long, gotSeqNr: Long)
+    extends DeadLetterSuppression
+  private[query] final case class ReplayFailed(cause: Throwable)
+    extends DeadLetterSuppression with NoSerializationVerificationNeeded
+
+}
+
+private[query] abstract class AbstractEventsByTagPublisher(
```

Comment:
In my opinion the events by tag publisher and fetcher are over complex, quite difficult to understand and reason about. I am pretty sure it could be simplified and share code with other publishers. But that could be done later separately since we have the tests.